### PR TITLE
feat(models): Mellum2-12B-A2.5B architecture port (Track G)

### PR DESCRIPTION
**Pull request by magicnight**
## Summary
Track G tested-tier: the **mellum** architecture (JetBrains Mellum2-12B-A2.5B — 28 layers alternating 21 sliding-window + 7 full-attention, 64-expert top-8 MoE), ported from mlx-lm's `mellum.py` with the full §1.2 fixture-parity methodology and registered via the zero-fork overlay.

- Per-layer RoPE: YaRN on full-attention layers (mscale formula verified byte-identical to the checkpoint's `attention_factor`), default RoPE on sliding layers; dual masks; mixed `KVCacheSimple`/`RotatingKVCache` cache.
- **MoE tie-break correctness**: router matches Python `argpartition(softmax)[-k:]` exactly — the sibling Qwen3MoE raw-logits pattern selects a different expert under an exact gate tie (caught by the parity fixture mid-port, fixed, and guarded with a do-not-simplify doc note).
- Existing platform gates cover the mixed-cache model with zero new code: batching refused (allowlist + dense converter), speculative refused once a sliding cache wraps, prompt-cache heldCount independent of rotating offsets — each gate individually verified in review.

## Review & verification
- Adversarial review (opus): **APPROVE** (0 CRITICAL/HIGH) — every numerical path (argpartition kth/slice, YaRN mscale, window-mask formula, weightedExpertSum axis, rope-offset ordering) verified against source; the one MEDIUM (stale docstring contradicting the tie-break fix) corrected pre-merge.
- Parity: 4 fixtures (mlx 0.32.0 + mlx-lm@e476a22, PEP 723 pins) all `allClose 1e-4` on real Metal, independently re-run.
- Real-weights smoke: 7.3GB 4-bit checkpoint, coherent 245-token thinking-stream generation, **68.9 tok/s** (independent re-run).
- Final full double-run: 452 tests / 56 suites green (xcodebuild + bare).

---

> [!NOTE]
> **Medium Risk**
> Large new inference path (MoE routing, YaRN, mixed caches, weight sanitize); risk is mitigated by extensive numerical parity tests and gated real-model smoke, but correctness bugs would affect generation quality for Mellum loads.
> 
> **Overview**
> Adds a pure-Swift **Mellum 2** (`model_type: mellum`) implementation and registers it through `ModelOverlay`, so the stock `LLMModelFactory` can load Mellum checkpoints without forking mlx-swift-lm.
> 
> The new `Mellum2Model` wires Qwen3-style attention (per-head q/k norm, layer-type RoPE: YaRN on full layers and default RoPE on sliding), a 64-expert sparse MoE on every layer (softmax top-k routing aligned with Python `argpartition`), alternating sliding-window vs full attention with matching masks, and a **mixed KV cache** (`RotatingKVCache` vs `KVCacheSimple`). Checkpoint `sanitize` stacks per-expert weights into `switch_mlp` when needed.
> 
> Verification adds mlx-lm parity fixtures (`capture_mellum2.py`), component and end-to-end `allClose` tests at 1e-4, overlay registration tests, and an opt-in real-weights smoke test through `MLXSwiftEngine`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4d197f3897f247d86e1a510a3fbe919496768887. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: MacMLXCore/Sources/MacMLXCore/Models/Mellum2.swift [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+
+    /// Mixed per-layer cache: `RotatingKVCache` for sliding layers, plain
+    /// `KVCacheSimple` for full-attention layers. Do NOT fall back to the
+    /// `KVCacheDimensionProvider` default (which is uniform `KVCacheSimple`).
+    func newCache(parameters: GenerateParameters?) -> [KVCache] {
+        config.layerTypes.map { type in
+            type == "full_attention"
+                ? KVCacheSimple()
+                : RotatingKVCache(maxSize: config.slidingWindow)
+        }
```
[the remark this answers is not in the document]
Comment (magicnight) [reply MacMLXCore/Sources/MacMLXCore/Models/Mellum2.swift:403]: Verified: partially — the missing count guard is real, but every access is bounds-checked with an explicit full_attention fallback (layerType lookup and the mask selection both guard on index < layerTypes.count), so there is no mis-typed cache or wrong-RoPE state on real configs; the residual risk is a malformed config silently falling back rather than erroring. Recalibrated High→Low; a config-validation guard (non-empty layerTypes must match numHiddenLayers) is tracked as a hardening follow-up.

File: MacMLXCore/Sources/MacMLXCore/Models/Mellum2.swift [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+                for e in 0 ..< config.numExperts {
+                    guard
+                        let w = weights.removeValue(
+                            forKey: "\(prefix).mlp.experts.\(e).\(n).weight")
+                    else { break }
+                    joined.append(w)
+                }
+                if joined.count == config.numExperts {
+                    weights["\(prefix).mlp.switch_mlp.\(n).weight"] = stacked(joined)
+                }
```
[the remark this answers is not in the document]
Comment (magicnight) [reply MacMLXCore/Sources/MacMLXCore/Models/Mellum2.swift:437]: Verified: confirmed — experts removed before the completeness check were silently lost on partial sets. Fixed in #83: sanitize now gathers non-destructively, verifies the full expert set, and only then removes+stacks; incomplete sets leave the checkpoint untouched. (Quantized per-expert scales/biases stacking is a separate pre-existing gap, unreachable today since shipped quantized conversions are pre-stacked; tracked.)
